File: legacy_php/planner_api.php
<?php

function current_user_token() {
  if (!empty($_SERVER['HTTP_X_UID'])) return $_SERVER['HTTP_X_UID'];
  if (!empty($_GET['uid'])) return $_GET['uid'];
  if (!empty($_COOKIE['uid'])) return $_COOKIE['uid'];
  return null;
}

// api.php — Travel Planner backend (table: trips + trip_points + points)

header('Content-Type: application/json; charset=utf-8');
header('Access-Control-Allow-Origin: *'); // 前後端不同 port/domain 需要 CORS
header('Access-Control-Allow-Methods: GET, POST, PUT, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type, Authorization, X-UID');
if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') 
  { http_response_code(204); exit; }

//若沒帶 path，回簡單的說明
if (!isset($_GET['path']) && empty($_SERVER['PATH_INFO'])) {
  send_json(['ok' => true, 'routes' => ['trips','trips/{id}','user_trips','user_trips/{id}','ai/legs']]);
}


// === DB 連線（依你的 XAMPP/MAMP 設定修改） ===
$DB_HOST = getenv('DB_HOST') ?: '127.0.0.1';
$DB_USER = getenv('DB_USER') ?: 'root';
$DB_PASS = getenv('DB_PASS') ?: '';              // XAMPP/MAMP 常見是空字串
$DB_NAME = getenv('DB_NAME') ?: 'travel_planner';
$DB_PORT = getenv('DB_PORT') ?: '3306';

try {
  $pdo = new PDO(
    "mysql:host=$DB_HOST;port=$DB_PORT;dbname=$DB_NAME;charset=utf8mb4",
    $DB_USER, $DB_PASS,
    [
      PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
      PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
    ]
  );
} catch (PDOException $e) {
  http_response_code(500);
  echo json_encode(['error' => 'DB connect failed', 'detail' => $e->getMessage()], JSON_UNESCAPED_UNICODE);
  exit;
}

// 解析路徑：支援兩種寫法：
// 1) /planner_api.php?path=user_trips/123
// 2) /planner_api.php/user_trips/123  （PHP 內建伺服器會放在 PATH_INFO）
$path = $_GET['path'] ?? '';
if ($path === '' && !empty($_SERVER['PATH_INFO'])) {
  $path = ltrim($_SERVER['PATH_INFO'], '/');
}
$parts = explode('/', trim($path, '/'));

function send_json($data, $code = 200){
  http_response_code($code);
  echo json_encode($data, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
  exit;
}

// ===== /trips（列表） =====
if (!empty($parts[0]) && $parts[0] === 'trips' && count($parts) === 1) {
  $stmt = $pdo->query("SELECT id, title FROM trips ORDER BY id");
  $rows = $stmt->fetchAll();
  send_json(['items' => array_map(fn($r) => [
    'id'    => (int)$r['id'],
    'title' => $r['title'],
  ], $rows)]);
}

// GET /trips/{id} 取得行程與點位（依 seq 排序）
if (count($parts) === 2 && $parts[0] === 'trips') {
  $tripId = (int)$parts[1];

  // 1) 取行程
  $stmt = $pdo->prepare('SELECT id, title FROM trips WHERE id = ?');
  $stmt->execute([$tripId]);
  $trip = $stmt->fetch();
  if (!$trip) send_json(['error' => 'Trip not found'], 404);

  // 2) 取行程的點位（join trip_points → points，依 seq）
  $sql = "SELECT p.name, p.lat, p.lng
          FROM trip_points tp
          JOIN points p ON p.id = tp.point_id
          WHERE tp.trip_id = ?
          ORDER BY tp.seq";
  $stmt = $pdo->prepare($sql);
  $stmt->execute([$tripId]);
  $rows = $stmt->fetchAll();

  $points = array_map(fn($r) => [
    'name' => $r['name'],
    'lat'  => (float)$r['lat'],
    'lng'  => (float)$r['lng'],
  ], $rows);

  send_json([
    'id'     => (int)$trip['id'],
    'title'  => $trip['title'],
    'points' => $points,
  ]);
}

// === POST /trips ===
if ($_SERVER['REQUEST_METHOD'] === 'POST' && preg_match('#^trips$#', $path)) {
    $input = json_decode(file_get_contents('php://input'), true);
    $title = $input['title'] ?? '未命名行程';
    $country = $input['country'] ?? 'Unknown';
    $region = $input['region'] ?? '';
    $notes = $input['notes'] ?? '';
    $itinerary = implode(',', array_map(fn($p)=>$p['name'], $input['points'] ?? []));

    $stmt = $pdo->prepare("INSERT INTO trips (title, country, region, notes, itinerary, source, curation_status) VALUES (?,?,?,?,?,'manual','published')");
    $stmt->execute([$title, $country, $region, $notes, $itinerary]);
    $tripId = $pdo->lastInsertId();

    // 同步儲存 trip_points
    $points = $input['points'] ?? [];
    $seq = 1;
    foreach ($points as $p) {
        $pid = null;
        // 若資料庫已有此點，則查找 id
        if (isset($p['lat'], $p['lng'])) {
            $stmt2 = $pdo->prepare("SELECT id FROM points WHERE ABS(lat-?)<0.0001 AND ABS(lng-?)<0.0001 LIMIT 1");
            $stmt2->execute([$p['lat'], $p['lng']]);
            $pid = $stmt2->fetchColumn();
        }
        // 若沒有則新建
        if (!$pid) {
            $stmt3 = $pdo->prepare("INSERT INTO points (name, country, region, city, lat, lng) VALUES (?,?,?,?,?,?)");
            $stmt3->execute([$p['name'] ?? '未命名', $country, $region, '', $p['lat'], $p['lng']]);
            $pid = $pdo->lastInsertId();
        }
        $stmt4 = $pdo->prepare("INSERT INTO trip_points (trip_id, point_id, seq) VALUES (?,?,?)");
        $stmt4->execute([$tripId, $pid, $seq++]);
    }

    echo json_encode(['success'=>true, 'trip_id'=>$tripId]);
    exit;
}

//建立顧客行程
if ($_SERVER['REQUEST_METHOD']==='POST' && preg_match('#^user_trips$#', $path)) {
  $token = current_user_token();
  if (!$token) send_json(['error'=>'missing user_token'], 400);

  $body = json_decode(file_get_contents('php://input'), true) ?: [];
  $title = trim($body['title'] ?? '未命名行程');
  $notes = $body['notes'] ?? null;
  $base  = isset($body['base_trip_id']) ? (int)$body['base_trip_id'] : null;
  $pts   = is_array($body['points'] ?? null) ? $body['points'] : [];
  if (!$pts) send_json(['error'=>'points required'], 400);

  $pdo->prepare("INSERT INTO user_trips (user_id,user_token,title,notes,base_trip_id) VALUES (NULL,?,?,?,?)")
      ->execute([$token,$title,$notes,$base]);
  $id = (int)$pdo->lastInsertId();

  $seq = 1;
  $stmt = $pdo->prepare("INSERT INTO user_trip_points (user_trip_id,seq,name,lat,lng,place_id,meta_json)
                         VALUES (?,?,?,?,?,?,?)");
  foreach ($pts as $p) {
    $name  = trim($p['name'] ?? '未命名');
    $lat   = (float)($p['lat'] ?? 0);
    $lng   = (float)($p['lng'] ?? 0);
    $place = $p['placeId'] ?? null;

    $pp = $p;
    if (!isset($pp['seq'])) $pp['seq'] = $seq;
    if (!isset($pp['day'])) $pp['day'] = $pp['day'] ?? 1;
    $meta = build_meta_json_from_point($pp);

    $stmt->execute([$id, $seq++, $name, $lat, $lng, $place, $meta]);
  }
  send_json(['success'=>true, 'user_trip_id'=>$id], 201);
}

//列出自己的行程
if ($_SERVER['REQUEST_METHOD']==='GET' && preg_match('#^user_trips$#', $path)) {
  $token = current_user_token();
  if (!$token) send_json(['items'=>[]]);
  $q = $pdo->prepare("SELECT id,title,created_at,updated_at FROM user_trips WHERE user_token=? ORDER BY updated_at DESC");
  $q->execute([$token]);
  send_json(['items'=>$q->fetchAll(PDO::FETCH_ASSOC)]);
}

//讀取單一行程
if ($_SERVER['REQUEST_METHOD']==='GET' && preg_match('#^user_trips/(\d+)$#', $path,$m)) {
  $id = (int)$m[1];
  $q = $pdo->prepare("SELECT id,user_token,title,notes,base_trip_id FROM user_trips WHERE id=?");
  $q->execute([$id]);
  $trip = $q->fetch(PDO::FETCH_ASSOC);
  if (!$trip) send_json(['error'=>'not found'],404);

  $token = current_user_token();
  if ($token && $trip['user_token'] !== $token) send_json(['error'=>'forbidden'],403);

  $p = $pdo->prepare("SELECT seq,name,lat,lng,place_id,meta_json FROM user_trip_points WHERE user_trip_id=? ORDER BY seq");
  $p->execute([$id]);

  $points = $p->fetchAll(PDO::FETCH_ASSOC); // 注意：是 PDO::FETCH_ASSOC，中間是底線

  // 先把資料放到變數（用 array()，避免隱藏字元/短陣列解析問題）
  $response = array(
    'trip'   => $trip,
    'points' => $points
  );

  send_json($response);
}

// === POST /ai/legs ：根據 stops 推薦每段交通模式 ===
if ($_SERVER['REQUEST_METHOD']==='POST' && preg_match('#^ai/legs$#', $path)) {
  $body = json_decode(file_get_contents('php://input'), true) ?: [];
  $stops = $body['stops'] ?? [];
  $prefs = $body['prefs'] ?? [];
  $day   = $body['day'] ?? 1;

  // --- 1) 建立系統/使用者提示與 JSON Schema（你之後可接真實 LLM） ---
  $schema = [
    'type'=>'object',
    'properties'=>[
      'legs'=>[
        'type'=>'array',
        'items'=>[
          'type'=>'object',
          'properties'=>[
            'mode'=>['type'=>'string'], // DRIVING/WALKING/BICYCLING/TRANSIT/FLIGHT/TRAIN/BUS/FERRY/MANUAL
            'departAt'=>['type'=>['string','null']], // ISO 8601
            'manualDistanceKm'=>['type'=>['number','null']],
            'manualDurationMin'=>['type'=>['number','null']],
            'bufferMin'=>['type'=>['number','null']],
            'cruiseKmh'=>['type'=>['number','null']],
            'notes'=>['type'=>['string','null']]
          ],
          'required'=>['mode']
        ]
      ]
    ],
    'required'=>['legs']
  ];

  // --- 2) 這裡你可呼叫真實 LLM。現階段先用簡單 heuristic 產生假結果讓前端串起來 ---
  $legs = [];
  for ($i=1; $i<count($stops); $i++){
    $A = $stops[$i-1]; $B = $stops[$i];
    $dKm = 2 * 6371 * asin(sqrt(
      pow(sin(deg2rad(($B['lat']-$A['lat'])/2)),2) +
      cos(deg2rad($A['lat'])) * cos(deg2rad($B['lat'])) *
      pow(sin(deg2rad(($B['lng']-$A['lng'])/2)),2)
    ));
    // 極簡規則：<1km 走路、<5km 單車、<200km 開車、>=200km 飛機（僅示範）
    if ($dKm < 1) {
      $legs[] = ['mode'=>'WALKING','notes'=>"距離約 ".round($dKm,1)."km，建議步行"];
    } else if ($dKm < 5) {
      $legs[] = ['mode'=>'BICYCLING','notes'=>"距離約 ".round($dKm,1)."km，建議單車"];
    } else if ($dKm < 200) {
      $legs[] = ['mode'=>'DRIVING','notes'=>"距離約 ".round($dKm,0)."km，建議開車/計程車"];
    } else {
      $legs[] = ['mode'=>'FLIGHT','notes'=>"距離約 ".round($dKm,0)."km，建議飛機（直線估算）"];
    }
  }

  // 回傳符合 schema 的物件
  send_json(['legs'=>$legs]);
}

//覆寫更新
if ($_SERVER['REQUEST_METHOD']==='PUT' && preg_match('#^user_trips/(\d+)$#',$path,$m)) {
  $id = (int)$m[1];
  $token = current_user_token();
  if (!$token) send_json(['error'=>'missing user_token'],400);

  $q=$pdo->prepare("SELECT user_token FROM user_trips WHERE id=?");
  $q->execute([$id]);
  $owner=$q->fetchColumn();
  if(!$owner) send_json(['error'=>'not found'],404);
  if($owner!==$token) send_json(['error'=>'forbidden'],403);

  $body = json_decode(file_get_contents('php://input'), true) ?: [];
  if (isset($body['title'])) $pdo->prepare("UPDATE user_trips SET title=?,updated_at=NOW() WHERE id=?")->execute([trim($body['title']),$id]);
  if (array_key_exists('notes',$body)) $pdo->prepare("UPDATE user_trips SET notes=?,updated_at=NOW() WHERE id=?")->execute([$body['notes'],$id]);

  if (isset($body['points']) && is_array($body['points'])) {
    $pdo->prepare("DELETE FROM user_trip_points WHERE user_trip_id=?")->execute([$id]);
    $stmt=$pdo->prepare("INSERT INTO user_trip_points (user_trip_id,seq,name,lat,lng,place_id,meta_json) VALUES (?,?,?,?,?,?,?)");
    $seq=1;
    foreach ($body['points'] as $p) {
      $name  = trim($p['name'] ?? '未命名');
      $lat   = (float)($p['lat'] ?? 0);
      $lng   = (float)($p['lng'] ?? 0);
      $place = $p['placeId'] ?? null;

      // ✅ 先補 seq/day 再組 meta_json（與 POST 同步）
      $pp = $p;
      if (!isset($pp['seq'])) $pp['seq'] = $seq;
      if (!isset($pp['day'])) $pp['day'] = $pp['day'] ?? 1;
      $meta = build_meta_json_from_point($pp);

      $stmt->execute([$id,$seq++,$name,$lat,$lng,$place,$meta]);
    }
  }
  send_json(['success'=>true,'user_trip_id'=>$id]);
}

// 其他路徑
send_json(['error' => 'Not found'], 404);


// 將前端的點資料組成 meta_json（合併：既有欄位 + leg + stayMin + day + seq + 客戶端 meta_json）
function build_meta_json_from_point($p){
  // 你原本就存的欄位
  $meta = [
    'rating'           => $p['rating']           ?? null,
    'userRatingsTotal' => $p['userRatingsTotal'] ?? null,
    'photoUrl'         => $p['photoUrl']         ?? null,
    'badges'           => $p['badges']           ?? [],
  ];

  // 交通與停留資訊
  if (array_key_exists('leg', $p))     $meta['leg']     = $p['leg'];
  if (array_key_exists('stayMin', $p)) $meta['stayMin'] = $p['stayMin'];

  // ✅ 加上分日與順序（讓前端回讀能還原）
  if (isset($p['day'])) $meta['day'] = (int)$p['day'];
  if (isset($p['seq'])) $meta['seq'] = (int)$p['seq'];

  // 若前端已經有組好的 meta_json（字串或陣列），就跟上面合併（前端優先）
  if (isset($p['meta_json'])) {
    $fromClient = null;
    if (is_string($p['meta_json'])) {
      $tmp = json_decode($p['meta_json'], true);
      if (is_array($tmp)) $fromClient = $tmp;
    } elseif (is_array($p['meta_json'])) {
      $fromClient = $p['meta_json'];
    }
    if (is_array($fromClient)) {
      // 用客戶端的值覆蓋（例如客戶端自己帶的 day/seq/leg/stayMin）
      $meta = array_replace($meta, $fromClient);
    }
  }

  return json_encode($meta, JSON_UNESCAPED_UNICODE);
}

File: scripts/planner_api.py
# -*- coding: utf-8 -*-
import os, json, math
from datetime import datetime
from flask import (
    Flask, request, make_response, jsonify,
    render_template
)
import pymysql

# === Resolve project root from this file (scripts/planner_api.py) ===
SCRIPT_DIR    = os.path.abspath(os.path.dirname(__file__))        # .../Project8/scripts
PROJECT_ROOT  = os.path.abspath(os.path.join(SCRIPT_DIR, ".."))   # .../Project8
TEMPLATES_DIR = os.path.join(PROJECT_ROOT, "templates")           # .../Project8/templates
STATIC_DIR    = os.path.join(PROJECT_ROOT, "static")              # .../Project8/static

app = Flask(
    __name__,
    template_folder=TEMPLATES_DIR,   # /templates (e.g. map.html)
    static_folder=STATIC_DIR,        # /static   (css/js)
    static_url_path="/static"
)

# ===== CORS (match PHP behavior, but DO NOT force JSON for non-API) =====
ALLOWED_HEADERS = "Content-Type, Authorization, X-UID"

@app.after_request
def add_cors_headers(resp):
    resp.headers["Access-Control-Allow-Origin"]  = "*"
    resp.headers["Access-Control-Allow-Methods"] = "GET, POST, PUT, OPTIONS"
    resp.headers["Access-Control-Allow-Headers"] = ALLOWED_HEADERS
    # Only set JSON content-type for API paths. Let HTML/CSS/JS keep defaults.
    if request.path.startswith(("/planner/", "/planner_api.", "/api.py")):
        resp.headers["Content-Type"] = "application/json; charset=utf-8"
    return resp

# ====== FRONTEND PAGES ======
@app.route("/")
def home():
    return render_template("map.html")

@app.route("/map")
def map_page():
    return render_template("map.html")

# (Optional) quick path debugger
@app.route("/_debug/paths")
def _debug_paths():
    return {
        "SCRIPT_DIR": SCRIPT_DIR,
        "PROJECT_ROOT": PROJECT_ROOT,
        "TEMPLATES_DIR": TEMPLATES_DIR,
        "STATIC_DIR": STATIC_DIR,
    }

# ---------- Shared dispatcher so both API URL styles work ----------
def _dispatch_by_path(path: str):
    path = (path or "").strip().strip("/")
    if not path:
        return json_ok({"ok": True, "routes": ["trips", "trips/{id}", "user_trips", "user_trips/{id}", "ai/legs"]})

    try:
        if path == "trips" and request.method == "GET":
            return list_trips()

        if path.startswith("trips/") and request.method == "GET":
            trip_id = int(path.split("/", 1)[1])
            return get_trip(trip_id)

        if path == "trips" and request.method == "POST":
            return create_trip()

        if path == "user_trips" and request.method == "POST":
            return create_user_trip()

        if path == "user_trips" and request.method == "GET":
            return list_user_trips()

        if path.startswith("user_trips/") and request.method == "GET":
            ut_id = int(path.split("/", 1)[1])
            return get_user_trip(ut_id)

        if path.startswith("user_trips/") and request.method == "PUT":
            ut_id = int(path.split("/", 1)[1])
            return update_user_trip(ut_id)

        if path == "ai/legs" and request.method == "POST":
            return ai_legs()

        return json_err("Not found", 404)
    except ValueError:
        return json_err("Bad request", 400)
    except Exception as e:
        return json_err(f"Server error: {str(e)}", 500)

# ---------- Query-style endpoints (…?path=trips) ----------
@app.route("/planner/planner_api.py", methods=["GET", "POST", "PUT", "OPTIONS"])
@app.route("/planner/planner_api.php", methods=["GET", "POST", "PUT", "OPTIONS"])  # backward-compatible
@app.route("/api.py", methods=["GET", "POST", "PUT", "OPTIONS"])                   # optional
def api_entry():
    if request.method == "OPTIONS":
        return ("", 204)
    path = (request.args.get("path") or "").strip().strip("/")
    return _dispatch_by_path(path)

# ---------- Slash-style endpoints (/planner/planner_api.php/trips) ----------
@app.route("/planner/planner_api.py/<path:subpath>", methods=["GET", "POST", "PUT", "OPTIONS"])
@app.route("/planner/planner_api.php/<path:subpath>", methods=["GET", "POST", "PUT", "OPTIONS"])
def api_entry_slash(subpath):
    if request.method == "OPTIONS":
        return ("", 204)
    return _dispatch_by_path(subpath)

# ---------- Root-level query-style endpoints (/planner_api.php?path=...) ----------
@app.route("/planner_api.php", methods=["GET","POST","PUT","OPTIONS"])
@app.route("/planner_api.py",  methods=["GET","POST","PUT","OPTIONS"])
def api_entry_root():
    if request.method == "OPTIONS":
        return ("", 204)
    path = (request.args.get("path") or "").strip().strip("/")
    return _dispatch_by_path(path)

# ---------- Root-level slash-style endpoints (/planner_api.php/trips) ----------
@app.route("/planner_api.php/<path:subpath>", methods=["GET","POST","PUT","OPTIONS"])
@app.route("/planner_api.py/<path:subpath>",  methods=["GET","POST","PUT","OPTIONS"])
def api_entry_root_slash(subpath):
    if request.method == "OPTIONS":
        return ("", 204)
    return _dispatch_by_path(subpath)


# ====== DB connection ======
def get_db():
    return pymysql.connect(
        host=os.getenv("DB_HOST", "127.0.0.1"),
        port=int(os.getenv("DB_PORT", "3306")),
        user=os.getenv("DB_USER", "root"),
        password=os.getenv("DB_PASS", ""),
        database=os.getenv("DB_NAME", "travel_planner"),
        charset="utf8mb4",
        cursorclass=pymysql.cursors.DictCursor,
        autocommit=True,
    )

# ====== Auth token (X-UID / uid) ======
def current_user_token():
    token = request.headers.get("X-UID") or request.args.get("uid") or request.cookies.get("uid")
    return token

# ====== Helpers ======
def json_ok(data, code=200):
    return make_response(json.dumps(data, ensure_ascii=False, default=str), code)

def json_err(msg, code=400):
    return json_ok({"error": msg}, code)

def row_map_point(r):
    return {"name": r.get("name"), "lat": float(r.get("lat")), "lng": float(r.get("lng"))}

def build_meta_json_from_point(p: dict) -> str:
    meta = {
        "rating":           p.get("rating"),
        "userRatingsTotal": p.get("userRatingsTotal"),
        "photoUrl":         p.get("photoUrl"),
        "badges":           p.get("badges", []),
    }
    if "leg" in p:     meta["leg"] = p["leg"]
    if "stayMin" in p: meta["stayMin"] = p["stayMin"]

    # ✅ 補上分日與順序（很重要）
    if "day" in p: meta["day"] = int(p["day"])
    if "seq" in p: meta["seq"] = int(p["seq"])


    # 與客戶端傳來的 meta_json 合併（保留客製欄位）
    if "meta_json" in p:
        client = None
        mj = p["meta_json"]
        if isinstance(mj, str):
            try: client = json.loads(mj)
            except Exception: client = None
        elif isinstance(mj, dict):
            client = mj
        if isinstance(client, dict):
            meta.update(client)

    return json.dumps(meta, ensure_ascii=False)


def haversine_km(a, b):
    to_rad = lambda x: (x * math.pi / 180.0)
    R = 6371.0
    dlat = to_rad(b["lat"] - a["lat"])
    dlng = to_rad(b["lng"] - a["lng"])
    la1, la2 = to_rad(a["lat"]), to_rad(b["lat"])
    h = math.sin(dlat/2)**2 + math.cos(la1)*math.cos(la2)*math.sin(dlng/2)**2
    return 2 * R * math.asin(math.sqrt(h))

def guess_cruise_kmh(mode):
    return 120 if mode == "TRAIN" else 60 if mode == "BUS" else 35 if mode == "FERRY" else 800

def guess_buffer_min(mode):
    return 90 if mode == "FLIGHT" else 20 if mode == "TRAIN" else 10

# ===== trips =====
def list_trips():
    with get_db() as conn, conn.cursor() as cur:
        cur.execute("SELECT id, title FROM trips ORDER BY id")
        rows = cur.fetchall()
        items = [{"id": int(r["id"]), "title": r["title"]} for r in rows]
    return json_ok({"items": items})

def get_trip(trip_id: int):
    with get_db() as conn, conn.cursor() as cur:
        cur.execute("SELECT id, title FROM trips WHERE id=%s", (trip_id,))
        trip = cur.fetchone()
        if not trip:
            return json_err("Trip not found", 404)

        sql = """
        SELECT p.name, p.lat, p.lng
        FROM trip_points tp
        JOIN points p ON p.id = tp.point_id
        WHERE tp.trip_id = %s
        ORDER BY tp.seq
        """
        cur.execute(sql, (trip_id,))
        rows = cur.fetchall()
        points = [row_map_point(r) for r in rows]

    return json_ok({"id": int(trip["id"]), "title": trip["title"], "points": points})

def create_trip():
    body = request.get_json(silent=True) or {}
    title   = body.get("title", "Untitled Trip")
    country = body.get("country", "Unknown")
    region  = body.get("region", "")
    notes   = body.get("notes", "")
    pts     = body.get("points", []) or []
    itinerary = ",".join([str(p.get("name") or "") for p in pts])

    with get_db() as conn, conn.cursor() as cur:
        cur.execute(
            "INSERT INTO trips (title, country, region, notes, itinerary, source, curation_status) VALUES (%s,%s,%s,%s,%s,'manual','published')",
            (title, country, region, notes, itinerary)
        )
        trip_id = cur.lastrowid

        seq = 1
        for p in pts:
            pid = None
            if ("lat" in p) and ("lng" in p):
                cur.execute(
                    "SELECT id FROM points WHERE ABS(lat-%s)<0.0001 AND ABS(lng-%s)<0.0001 LIMIT 1",
                    (p["lat"], p["lng"])
                )
                row = cur.fetchone()
                pid = row["id"] if row else None

            if not pid:
                cur.execute(
                    "INSERT INTO points (name, country, region, city, lat, lng) VALUES (%s,%s,%s,%s,%s,%s)",
                    (p.get("name") or "Untitled", country, region, "", p.get("lat"), p.get("lng"))
                )
                pid = cur.lastrowid

            cur.execute(
                "INSERT INTO trip_points (trip_id, point_id, seq) VALUES (%s,%s,%s)",
                (trip_id, pid, seq)
            )
            seq += 1

    return json_ok({"success": True, "trip_id": int(trip_id)})

# ===== user_trips =====
def create_user_trip():
    token = current_user_token()
    if not token:
        return json_err("missing user_token", 400)

    body  = request.get_json(silent=True) or {}
    title = (body.get("title") or "Untitled Trip").strip()
    notes = body.get("notes")
    base  = int(body.get("base_trip_id")) if body.get("base_trip_id") is not None else None
    pts   = body.get("points") or []
    if not pts:
        return json_err("points required", 400)

    with get_db() as conn, conn.cursor() as cur:
        cur.execute(
            "INSERT INTO user_trips (user_id, user_token, title, notes, base_trip_id) VALUES (NULL,%s,%s,%s,%s)",
            (token, title, notes, base)
        )
        uid = int(cur.lastrowid)

        sql = """INSERT INTO user_trip_points (user_trip_id, seq, name, lat, lng, place_id, meta_json)
                 VALUES (%s,%s,%s,%s,%s,%s,%s)"""
        seq = 1
        for p in pts:
            name  = (p.get("name") or "Untitled").strip()
            lat   = float(p.get("lat") or 0)
            lng   = float(p.get("lng") or 0)
            place = p.get("placeId")
            meta  = build_meta_json_from_point(p)
            cur.execute(sql, (uid, seq, name, lat, lng, place, meta))
            seq += 1

    return json_ok({"success": True, "user_trip_id": uid}, 201)

def list_user_trips():
    token = current_user_token()
    if not token:
        return json_ok({"items": []})
    with get_db() as conn, conn.cursor() as cur:
        cur.execute(
            "SELECT id,title,created_at,updated_at FROM user_trips WHERE user_token=%s ORDER BY updated_at DESC",
            (token,)
        )
        items = cur.fetchall()
    return json_ok({"items": items})

def get_user_trip(ut_id: int):
    with get_db() as conn, conn.cursor() as cur:
        cur.execute("SELECT id,user_token,title,notes,base_trip_id FROM user_trips WHERE id=%s", (ut_id,))
        trip = cur.fetchone()
        if not trip:
            return json_err("not found", 404)

        token = current_user_token()
        if token and trip["user_token"] != token:
            return json_err("forbidden", 403)

        cur.execute(
            "SELECT seq,name,lat,lng,place_id,meta_json FROM user_trip_points WHERE user_trip_id=%s ORDER BY seq",
            (ut_id,)
        )
        points = cur.fetchall()

    return json_ok({"trip": trip, "points": points})

def update_user_trip(ut_id: int):
    token = current_user_token()
    if not token:
        return json_err("missing user_token", 400)

    body = request.get_json(silent=True) or {}

    with get_db() as conn, conn.cursor() as cur:
        cur.execute("SELECT user_token FROM user_trips WHERE id=%s", (ut_id,))
        row = cur.fetchone()
        if not row:
            return json_err("not found", 404)
        if row["user_token"] != token:
            return json_err("forbidden", 403)

        if "title" in body:
            cur.execute("UPDATE user_trips SET title=%s, updated_at=NOW() WHERE id=%s", (body["title"].strip(), ut_id))
        if "notes" in body:
            cur.execute("UPDATE user_trips SET notes=%s, updated_at=NOW() WHERE id=%s", (body["notes"], ut_id))

        if isinstance(body.get("points"), list):
            cur.execute("DELETE FROM user_trip_points WHERE user_trip_id=%s", (ut_id,))
            sql = """INSERT INTO user_trip_points (user_trip_id, seq, name, lat, lng, place_id, meta_json)
                     VALUES (%s,%s,%s,%s,%s,%s,%s)"""
            seq = 1
            for p in body["points"]:
                name  = (p.get("name") or "Untitled").strip()
                lat   = float(p.get("lat") or 0)
                lng   = float(p.get("lng") or 0)
                place = p.get("placeId")
                meta  = build_meta_json_from_point(p)
                cur.execute(sql, (ut_id, seq, name, lat, lng, place, meta))
                seq += 1

    return json_ok({"success": True, "user_trip_id": ut_id})

# ===== AI/legs (heuristic identical to the PHP version) =====
def ai_legs():
    body  = request.get_json(silent=True) or {}
    stops = body.get("stops") or []
    legs = []
    for i in range(1, len(stops)):
        A, B = stops[i-1], stops[i]
        dkm = haversine_km({"lat": A["lat"], "lng": A["lng"]},
                           {"lat": B["lat"], "lng": B["lng"]})
        if dkm < 1:
            legs.append({"mode": "WALKING", "notes": f"Distance ~ {round(dkm,1)} km — walking recommended"})
        elif dkm < 5:
            legs.append({"mode": "BICYCLING", "notes": f"Distance ~ {round(dkm,1)} km — bicycling recommended"})
        elif dkm < 200:
            legs.append({"mode": "DRIVING", "notes": f"Distance ~ {round(dkm,0)} km — car/taxi recommended"})
        else:
            legs.append({"mode": "FLIGHT", "notes": f"Distance ~ {round(dkm,0)} km — flight recommended (great-circle estimate)"})
    return json_ok({"legs": legs})

# ===== Main (development) =====
if __name__ == "__main__":
    # Run from project root or anywhere:
    #   python scripts\planner_api.py
    port = int(os.getenv("PORT", "8000"))
    app.run(host="127.0.0.1", port=port, debug=True)
